From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Wed, 11 Feb 2026 03:36:54 +0100
Subject: ui/qrCode: Add a QR Code widget

The widget can only have a squared size that is picked using the maximum
between the provided width ad height properties, using the minimum of
the space available.

The qr code is always generated using the minimum sized image with a
transparent background (so that the widget background color is used)
while the foreground is colored following the theme.

The texture is then draw using the nearest filter so that it will adjust
to the actor size without the need to regenerate it on size changes
theme

Origin: https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/3212
---
 data/theme/gnome-shell-sass/_widgets.scss         |   1 +
 data/theme/gnome-shell-sass/widgets/_qr-code.scss |  14 +++
 data/theme/meson.build                            |   1 +
 js/js-resources.gresource.xml                     |   1 +
 js/misc/dependencies.js                           |   1 +
 js/ui/qrCode.js                                   | 119 ++++++++++++++++++++++
 6 files changed, 137 insertions(+)
 create mode 100644 data/theme/gnome-shell-sass/widgets/_qr-code.scss
 create mode 100644 js/ui/qrCode.js

diff --git a/data/theme/gnome-shell-sass/_widgets.scss b/data/theme/gnome-shell-sass/_widgets.scss
index 9eac62d..1e3fefa 100644
--- a/data/theme/gnome-shell-sass/_widgets.scss
+++ b/data/theme/gnome-shell-sass/_widgets.scss
@@ -45,5 +45,6 @@
 @import 'widgets/misc';
 @import 'widgets/keyboard';
 @import 'widgets/looking-glass';
+@import 'widgets/qr-code';
 // Lock / login screen
 @import 'widgets/login-lock';
diff --git a/data/theme/gnome-shell-sass/widgets/_qr-code.scss b/data/theme/gnome-shell-sass/widgets/_qr-code.scss
new file mode 100644
index 0000000..da04425
--- /dev/null
+++ b/data/theme/gnome-shell-sass/widgets/_qr-code.scss
@@ -0,0 +1,14 @@
+// QR Code
+.qr-code {
+  border-radius: $base_border_radius * .5;
+  border-width: 1em;
+  @if ($variant == 'light') {
+    $qrcode_bg_color: mix($fg_color, $bg_color, 8%);
+    background-color: $qrcode_bg_color;
+    border-color: $qrcode_bg_color;
+  } @else {
+    background-color: $system_fg_color;
+    border-color: $system_fg_color;
+    color: $system_bg_color;
+  }
+}
diff --git a/data/theme/meson.build b/data/theme/meson.build
index 8d01ae8..d6f2d8e 100644
--- a/data/theme/meson.build
+++ b/data/theme/meson.build
@@ -28,6 +28,7 @@ theme_sources = files([
   'gnome-shell-sass/widgets/_overview.scss',
   'gnome-shell-sass/widgets/_panel.scss',
   'gnome-shell-sass/widgets/_popovers.scss',
+  'gnome-shell-sass/widgets/_qr-code.scss',
   'gnome-shell-sass/widgets/_quick-settings.scss',
   'gnome-shell-sass/widgets/_screenshot.scss',
   'gnome-shell-sass/widgets/_scrollbars.scss',
diff --git a/js/js-resources.gresource.xml b/js/js-resources.gresource.xml
index b1c59f3..b8b20a6 100644
--- a/js/js-resources.gresource.xml
+++ b/js/js-resources.gresource.xml
@@ -103,6 +103,7 @@
     <file>ui/pointerWatcher.js</file>
     <file>ui/popupMenu.js</file>
     <file>ui/quickSettings.js</file>
+    <file>ui/qrCode.js</file>
     <file>ui/remoteSearch.js</file>
     <file>ui/ripples.js</file>
     <file>ui/runDialog.js</file>
diff --git a/js/misc/dependencies.js b/js/misc/dependencies.js
index f8f98d4..93f54c1 100644
--- a/js/misc/dependencies.js
+++ b/js/misc/dependencies.js
@@ -16,6 +16,7 @@ import 'gi://GioUnix?version=2.0';
 import 'gi://GDesktopEnums?version=3.0';
 import 'gi://GdkPixbuf?version=2.0';
 import 'gi://GnomeBG?version=4.0';
+import 'gi://GnomeQR?version=4.0';
 import 'gi://GnomeDesktop?version=4.0';
 import 'gi://Graphene?version=1.0';
 import 'gi://GWeather?version=4.0';
diff --git a/js/ui/qrCode.js b/js/ui/qrCode.js
new file mode 100644
index 0000000..5f2bb85
--- /dev/null
+++ b/js/ui/qrCode.js
@@ -0,0 +1,119 @@
+import Clutter from 'gi://Clutter';
+import Cogl from 'gi://Cogl';
+import GObject from 'gi://GObject';
+import Gio from 'gi://Gio';
+import GnomeQR from 'gi://GnomeQR';
+import St from 'gi://St';
+import {logErrorUnlessCancelled} from '../misc/errorUtils.js';
+
+Gio._promisify(GnomeQR, 'generate_qr_code_async');
+
+const QR_CODE_DEFAULT_SIZE = 150;
+const QR_CODE_TRANSPARENT_COLOR = new GnomeQR.Color({alpha: 0});
+
+export class QrCode extends St.Bin {
+    static [GObject.GTypeName] = 'QrCode';
+    static [GObject.properties] = {
+        'url': GObject.ParamSpec.string(
+            'url', null, null,
+            GObject.ParamFlags.READWRITE,
+            null),
+    };
+
+    static {
+        GObject.registerClass(this);
+    }
+
+    constructor(params) {
+        const qrSize = Math.max(params.width ?? 0, params.height ?? 0) ||
+            QR_CODE_DEFAULT_SIZE;
+
+        super({
+            styleClass: 'qr-code',
+            width: qrSize,
+            height: qrSize,
+            ...params,
+        });
+
+        this.set_child(new Clutter.Actor({
+            xExpand: true,
+            yExpand: true,
+            xAlign: Clutter.ActorAlign.FILL,
+            yAlign: Clutter.ActorAlign.FILL,
+            minificationFilter: Clutter.ScalingFilter.NEAREST,
+            magnificationFilter: Clutter.ScalingFilter.NEAREST,
+        }));
+
+        this._fgColor = null;
+
+        this.connect('notify::url', () =>
+            this._update().catch(logErrorUnlessCancelled));
+    };
+
+    vfunc_allocate(box) {
+        const width = box.get_width();
+        const height = box.get_height();
+
+        box.set_size(Math.min(width, height), Math.min(width, height));
+        super.vfunc_allocate(box);
+    }
+
+    vfunc_style_changed() {
+        super.vfunc_style_changed();
+
+        if (!this.get_parent())
+            return;
+
+        const node = this.get_theme_node();
+        const fgColor = node.get_foreground_color();
+
+        if (!this._fgColor?.equal(fgColor)) {
+            this._fgColor = fgColor;
+            this._update().catch(logErrorUnlessCancelled);
+        }
+    }
+
+    async _update() {
+        const fgColor = QrCode.#getGnomeQRColor(this._fgColor);
+
+        this._cancellable?.cancel();
+        const cancellable = new Gio.Cancellable();
+        this._cancellable = cancellable;
+
+        const [pixelData, qrSize] = await GnomeQR.generate_qr_code_async(
+            this.url,
+            0 /* size, are fine with the minimum value */,
+            QR_CODE_TRANSPARENT_COLOR,
+            fgColor,
+            GnomeQR.PixelFormat.RGBA_8888,
+            GnomeQR.EccLevel.LOW,
+            cancellable);
+
+        const coglContext = global.stage.context.get_backend().get_cogl_context();
+        const bpp = Cogl.pixel_format_get_bytes_per_pixel(Cogl.PixelFormat.RGBA_8888, 0);
+        const rowStride = qrSize * bpp;
+
+        const content = St.ImageContent.new_with_preferred_size(qrSize, qrSize);
+        content.set_bytes(
+            coglContext,
+            pixelData,
+            Cogl.PixelFormat.RGBA_8888,
+            qrSize,
+            qrSize,
+            rowStride);
+
+        this.child.set_content(content);
+    }
+
+    static #getGnomeQRColor(color) {
+        if (!color)
+            return null;
+
+        return new GnomeQR.Color({
+            red: color.red,
+            green: color.green,
+            blue: color.blue,
+            alpha: color.alpha,
+        });
+    }
+};
